В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [2]:
import numpy as np
from statsmodels.stats.weightstats import ztest

In [1]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [7]:
null_mean = 170
alpha = 0.05
print(f"Середнє вибірки: {np.mean(heights)}")
print(f"Стандартне відхилення вибірки: {np.std(heights)}")

Середнє вибірки: 170.6969696969697
Стандартне відхилення вибірки: 4.648133209409457


**Гіпотези**:
- Н0: середній ріст студентів = 170 см
- H1: середній ріст студентів > 170 см  




In [8]:
ztest_score, p_value = ztest(heights, value=null_mean, alternative='larger')

In [10]:
ztest_score, p_value

(np.float64(0.8482235371587619), np.float64(0.19815674571091746))

In [13]:
if p_value < alpha:
    print("Відхилити H0.")
else:
    print("H0 не може бути відхилена.")

H0 не може бути відхилена.


**Висновок:**

p-value = 0.19815 > 0.05, тому ми не відхиляємо H0. На рівні значущості 0.05 немає достатніх статистичних доказів, що середній ріст студентів у вибірці є істотно більшим за 170 см. Середнє вибірки (~170.7 см) лише незначно перевищує гіпотетичне середнє, і ця різниця може бути випадковою.

### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [15]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [17]:
print("Група A — середнє: {:.2f}, std: {:.2f}".format(
    np.mean(group_a_scores), np.std(group_a_scores)))
print("Група B — середнє: {:.2f}, std: {:.2f}".format(
    np.mean(group_b_scores), np.std(group_b_scores)))

Група A — середнє: 77.08, std: 4.31
Група B — середнє: 74.74, std: 4.46


**Гіпотези:**
- Н0: середні оцінки груп A і B однакові (різниця = 0)
- Н1: середня оцінка групи A більша за групу B - новий курс ефективніший

In [20]:
ztest_score, p_value = ztest(group_a_scores, group_b_scores, alternative='larger')

In [21]:
ztest_score, p_value

(np.float64(2.357354864751502), np.float64(0.009202824276264005))

In [22]:
if p_value < alpha:
    print("Відхилити H0.")
else:
    print("H0 не може бути відхилена.")

Відхилити H0.


**Висновок:**

Новий навчальний курс виявився ефективнішим. Це доказав z-тест, та середнє двох груп відчутно відрізняється.

**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

In [23]:
from scipy.stats import ttest_ind

In [28]:
group_a_small = group_a_scores[:5]

print("Група A: n={}, середнє: {:.2f}".format(len(group_a_small), np.mean(group_a_small)))
print("Група B: n={}, середнє: {:.2f}".format(len(group_b_scores), np.mean(group_b_scores)))

Група A: n=5, середнє: 76.44
Група B: n=40, середнє: 74.74


In [29]:
t_stat, p_value_small = ttest_ind(group_a_small, group_b_scores, alternative='greater')

In [30]:
t_stat, p_value_small

(np.float64(0.8167951030410814), np.float64(0.20927397387678665))

In [31]:
if p_value_small < alpha:
    print("Відхилити H0.")
else:
    print("H0 не може бути відхилена.")

H0 не може бути відхилена.


**Висновок:**

При зменшенні вибірки групи A до 5 спостережень p-value значно зросло і перевищило 0.05. Тому ми не можемо відхилити H0. З цього робимо висновок, що навіть якщо реальна різниця між групами існує, малий розмір вибірки не дозволяє її достовірно виявити.


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [32]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

In [34]:
print("До тренінгу  - середнє: {:.2f}, std: {:.2f}".format(np.mean(before_training), np.std(before_training)))
print("Після тренінгу - середнє: {:.2f}, std: {:.2f}".format(np.mean(after_training), np.std(after_training)))

До тренінгу  - середнє: 50.89, std: 8.58
Після тренінгу - середнє: 54.04, std: 8.96


**Гіпотези:**
- H0: тренінг не покращив показники - середня різниця (after - before) = 0
- H1: тренінг покращив показники - середня різниця > 0

In [36]:
differences = np.array(after_training) - np.array(before_training)

print("Середня різниця: {:.4f}".format(np.mean(differences)))
print("Std різниць: {:.4f}".format(np.std(differences)))

Середня різниця: 3.1500
Std різниць: 1.8562


In [39]:
differences

array([ 4.65,  3.03,  5.9 ,  2.47,  8.44,  4.26,  1.29,  0.86,  3.97,
        2.55,  4.42,  3.94,  2.85,  1.3 , -0.03,  2.11,  4.71,  3.42,
        0.51,  3.35,  3.77,  1.23,  3.31,  3.11,  0.72,  3.72,  4.12,
        5.17,  5.11,  0.24])

In [37]:
ztest_score, p_value = ztest(differences, value=0, alternative='larger')

In [38]:
ztest_score, p_value

(np.float64(9.13890009623916), np.float64(3.1545359919175505e-20))

In [40]:
if p_value < alpha:
    print("Відхилити H0.")
else:
    print("H0 не може бути відхилена.")

Відхилити H0.


**Висновок:**
p-value < 0.05, тому ми відхиляємо H0. На рівні значущості 0.05 є переконливі статистичні докази, що тренінг покращив продуктивність співробітників. Середня різниця (3.15 одиниці) є позитивною і статистично значущою.